src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java

```diff
@@ -0,0 +1,179 @@
+        }
+        return contentList;
+    }
+
+    private static List<Pdf417> plotStructuredAppendSymbols(List<String> contentList, Pdf417 template, boolean forceByteCompaction) {
+        int structuredAppendTotal = contentList.size();
+        List<Pdf417> symbols = new ArrayList<>(structuredAppendTotal);
+        for (int i = 0; i < structuredAppendTotal; i++) {
+            String content = contentList.get(i);
+            Pdf417 symbol = new Pdf417();
+            symbols.add(symbol);
+            clone(template, symbol);
+
+            symbol.setStructuredAppendPosition(i + 1);
+            symbol.setStructuredAppendTotal(structuredAppendTotal);
+
+            setContent(symbol, content, forceByteCompaction);
+        }
+        return symbols;
+    }
+
+    private static void clone(Pdf417 template, Pdf417 target) {
+        target.setFontName(template.getFontName());
+        target.setFontSize(template.getFontSize());
+        target.setDataType(template.getDataType());
+        target.setEmptyContentAllowed(template.getEmptyContentAllowed());
+        target.setHumanReadableAlignment(template.getHumanReadableAlignment());
+        target.setHumanReadableLocation(template.getHumanReadableLocation());
+        target.setModuleWidth(template.getModuleWidth());
+        target.setQuietZoneHorizontal(template.getQuietZoneHorizontal());
+        target.setQuietZoneVertical(template.getQuietZoneVertical());
+        target.setReaderInit(template.getReaderInit());
+        target.setDataColumns(template.getDataColumns());
+        target.setRows(template.getRows());
+        target.setPreferredEccLevel(template.getPreferredEccLevel());
+        target.setStructuredAppendFileId(template.getStructuredAppendFileId());
+        target.setStructuredAppendFileName(template.getStructuredAppendFileName());
+        target.setBarHeight(template.getBarHeight());
+        target.setMode(template.getMode());
+        target.setStructuredAppendIncludeSegmentCount(template.getStructuredAppendIncludeSegmentCount());
+    }
+
+    private static String setContentAndGetRemainingWithBinarySearch(String remainingContent, Pdf417 symbol, boolean forceByteCompaction) {
+        int low = 0;
+        int high = remainingContent.length();
+
+        while (low <= high) {
+            int mid = low + high >>> 1;
+            SetContentStatus result = trySetContent(remainingContent, symbol, mid, forceByteCompaction);
+            if (result == SetContentStatus.TRY_MORE) {
+                low = mid + 1;
+            } else {
+                if (result == SetContentStatus.MAX_REACHED) {
+                    return remainingContent.substring(mid);
+                }
+                high = mid - 1;
+            }
+        }
+
+        throw new RuntimeException("Failed to set content, this is a bug.");
+    }
+
+    private static SetContentStatus trySetContent(String remainingContent, Pdf417 symbol, int lengthToTry, boolean forceByteCompaction) {
```

Comment:
If I use the template instead of the derived symbol for testing max. content length, I could not override `plotSymbol` with an empty impl, which could lead in degraded performance if I remember correctly a past comment from your side. Or do I miss something here?

Reply:
Yes, you want to use a "lite" Pdf417 subclass, but you don't necessarily need to build the "lite" clone based on the template until you're in the `trySetContent` / `fits` method. My idea was that this simplifies the outer loop so that the combined outer + inner loop (discussed in a different comment) would be easier to understand (once combined) since the cloning noise is pushed down into `fits`.

Reply:
I was able to change this partially only: moved the symbol template handling into this method. But have not found a way yet to refactor this to a boolean returning method. I am quite sure there is such a way, but somehow I am struggling here right now. Do you mind creating a patch which shows your idea?

Reply:
I gave it another try, but was not able to find a solution which was nearly as performing as the current implementation unfortunately.
@gredler If you still think we can simplify this method, I would appreciate if you could give it a try.

Reply:
@gredler Gave it another try, and was able to implement it. Please let me know if that is what you were thinking of.